File: 349.intersection-of-two-arrays.cpp
/*
 * @lc app=leetcode id=349 lang=cpp
 *
 * [349] Intersection of Two Arrays
 *
 * https://leetcode.com/problems/intersection-of-two-arrays/description/
 *
 * algorithms
 * Easy (59.65%)
 * Likes:    787
 * Dislikes: 1188
 * Total Accepted:    354.3K
 * Total Submissions: 575.1K
 * Testcase Example:  '[1,2,2,1]\n[2,2]'
 *
 * Given two arrays, write a function to compute their intersection.
 * 
 * Example 1:
 * 
 * 
 * Input: nums1 = [1,2,2,1], nums2 = [2,2]
 * Output: [2]
 * 
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
 * Output: [9,4]
 * 
 * 
 * Note:
 * 
 * 
 * Each element in the result must be unique.
 * The result can be in any order.
 * 
 * 
 * 
 * 
 */

// @lc code=start
class Solution
{
public:
    vector<int> intersection(vector<int> &nums1, vector<int> &nums2)
    {
        map<int, bool> m;
        vector<int> res;
        for (auto p = nums1.begin(); p < nums1.end(); p++)
        {
            m[*p] = true;
        }

        for (auto p = nums2.begin(); p < nums2.end(); p++)
        {
            if (m[*p])
            {
                res.push_back(*p);
                m[*p] = false;
            }
        }

        return res;
    }
};
// @lc code=end

File: 349.intersection-of-two-arrays/349.intersection-of-two-arrays.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

/*
 * @lc app=leetcode id=349 lang=java
 *
 * [349] Intersection of Two Arrays
 *
 * https://leetcode.com/problems/intersection-of-two-arrays/description/
 *
 * algorithms
 * Easy (67.18%)
 * Likes:    1932
 * Dislikes: 1738
 * Total Accepted:    549.1K
 * Total Submissions: 817.3K
 * Testcase Example:  '[1,2,2,1]\n[2,2]'
 *
 * Given two integer arrays nums1 and nums2, return an array of their
 * intersection. Each element in the result must be unique and you may return
 * the result in any order.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: nums1 = [1,2,2,1], nums2 = [2,2]
 * Output: [2]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: nums1 = [4,9,5], nums2 = [9,4,9,8,4]
 * Output: [9,4]
 * Explanation: [4,9] is also accepted.
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= nums1.length, nums2.length <= 1000
 * 0 <= nums1[i], nums2[i] <= 1000
 * 
 * 
 */

// @lc code=start
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> s = new HashSet<>();
        for (int n : nums1) {
            s.add(n);
        }

        List<Integer> res = new ArrayList<>();
        for (int n : nums2) {
            if (s.contains(n)) {
                res.add(n);
                s.remove(n);
            }
        }

        return res.stream().mapToInt(i -> i).toArray();
    }
}
// @lc code=end

